.. include:: /properties.rst

..  codeberg-logo_special_fake2.svg; 'official' alternative to icon proposed by jakeg; 
    see https://codeberg.org/Codeberg/Community/issues/976
    derived from https://codeberg.org/Codeberg/Design/src/branch/main/logo/special/codeberg-logo_special_fake-transaprency.svg
    see https://codeberg.org/Codeberg/Design/issues/80#issuecomment-878178

.. |codebergicon| replace:: :raw-html:`<a class="svglink" href="https://codeberg.org/coalispr/" aria-label="Codeberg" target="_blank"><svg width="2em" height="2em" viewBox="0 0 4.233 4.233"><path fill="#aaa" d="M46.984 76.122a2.117 2.117 0 0 0-1.793 3.242l1.764-2.282c.013-.016.045-.016.058 0l.736.953h-.527l.011.042h.55l.155.2h-.648l.018.067h.68l.138.177h-.769l.024.085h.81l.123.158h-.889l.03.103h.939l.107.139h-1.008l.032.115h1.065l.099.128H47.56l.033.115h1.184a2.117 2.117 0 0 0-1.793-3.242zm.645 3.37.032.114h.94a2.24 2.24 0 0 0 .09-.114zm.068.243.031.114h.629a2.54 2.54 0 0 0 .125-.114zm.067.242.032.115h.212c.063-.036.121-.073.184-.115z" style="paint-order:markers fill stroke" transform="translate(-44.867 -75.991)"/><desc>"Codeberg icon adapted by brobr,see https://codeberg.org/Codeberg/Community/issues/976"</desc></svg></a>`


..  creative commons-logos 'FaCreativeCommons', 'FaCreativeCommonsBy' 
    from https://react-icons.github.io/react-icons/search?q=creative

.. |ccby4| replace:: :raw-html:`<a class="svglink" href="https://creativecommons.org/licenses/by/4.0/"  aria-label="CC BY 4.0" target="_blank"><svg stroke="currentColor" fill="#aaa" stroke-width="0" viewBox="0 0 496 512" height="24" width="24"><path d="M245.83 214.87l-33.22 17.28c-9.43-19.58-25.24-19.93-27.46-19.93-22.13 0-33.22 14.61-33.22 43.84 0 23.57 9.21 43.84 33.22 43.84 14.47 0 24.65-7.09 30.57-21.26l30.55 15.5c-6.17 11.51-25.69 38.98-65.1 38.98-22.6 0-73.96-10.32-73.96-77.05 0-58.69 43-77.06 72.63-77.06 30.72-.01 52.7 11.95 65.99 35.86zm143.05 0l-32.78 17.28c-9.5-19.77-25.72-19.93-27.9-19.93-22.14 0-33.22 14.61-33.22 43.84 0 23.55 9.23 43.84 33.22 43.84 14.45 0 24.65-7.09 30.54-21.26l31 15.5c-2.1 3.75-21.39 38.98-65.09 38.98-22.69 0-73.96-9.87-73.96-77.05 0-58.67 42.97-77.06 72.63-77.06 30.71-.01 52.58 11.95 65.56 35.86zM247.56 8.05C104.74 8.05 0 123.11 0 256.05c0 138.49 113.6 248 247.56 248 129.93 0 248.44-100.87 248.44-248 0-137.87-106.62-248-248.44-248zm.87 450.81c-112.54 0-203.7-93.04-203.7-202.81 0-105.42 85.43-203.27 203.72-203.27 112.53 0 202.82 89.46 202.82 203.26-.01 121.69-99.68 202.82-202.84 202.82z"></path></svg><svg stroke="currentColor" fill="#aaa" stroke-width="0" viewBox="0 0 496 512" height="24" width="24"><path d="M314.9 194.4v101.4h-28.3v120.5h-77.1V295.9h-28.3V194.4c0-4.4 1.6-8.2 4.6-11.3 3.1-3.1 6.9-4.7 11.3-4.7H299c4.1 0 7.8 1.6 11.1 4.7 3.1 3.2 4.8 6.9 4.8 11.3zm-101.5-63.7c0-23.3 11.5-35 34.5-35s34.5 11.7 34.5 35c0 23-11.5 34.5-34.5 34.5s-34.5-11.5-34.5-34.5zM247.6 8C389.4 8 496 118.1 496 256c0 147.1-118.5 248-248.4 248C113.6 504 0 394.5 0 256 0 123.1 104.7 8 247.6 8zm.8 44.7C130.2 52.7 44.7 150.6 44.7 256c0 109.8 91.2 202.8 203.7 202.8 103.2 0 202.8-81.1 202.8-202.8.1-113.8-90.2-203.3-202.8-203.3z"></path><desc>"Licence CC BY 4.0, Attribution 4.0 International: free to reuse with crediting the maker"</desc></svg></a>`


.. EUPL-logo for use as button; adapted by author.
   from https://upload.wikimedia.org/wikipedia/commons/9/91/Logo_EUPL.svg


.. |euplsvg| replace:: :raw-html:`<a class="svglink" href="https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12"  aria-label="EUPL-1.2" target="_blank"><svg width="3em" height="3em" viewBox="0 0 245.73 189.77"><defs><radialGradient id="Dégradé_sans_nom_8" cx="128" cy="94.88" r="76" gradientUnits="userSpaceOnUse"><stop offset="0" stop-color="#0178bc"/><stop offset="1" stop-color="#005894"/></radialGradient><style>.cls-1{fill:#dbdbdb}.cls-9{fill:#ffd941}</style></defs><path d="M43.73 133.08C26 131.88 0 118.73 0 118.73V92.15l37 15.73Z" class="cls-1"/><path d="M35.64 122.87C0 104 0 100 0 100v34.85A95.81 95.81 0 0 0 38 149" style="fill:#00569c" transform="translate(0 -33.12)"/><path d="M55.94 152.88c-11.94 0-32.94-3-32.94-3l5 16s18.37 5.61 41.68.8M219.35 80.28c13.65.6 20.65 2.6 20.65 2.6l5.57 18.85s-14.57-4.85-25.09-4.85" class="cls-1"/><path d="M220.24 121c15.76 8 25.49 24.25 25.49 24.25L238 167l-20.53-15.44" style="fill:#a8acad" transform="translate(0 -33.12)"/><path d="M218.88 145.29C236 158 238 167 238 167l-15 39s-12-19-18.49-26c0 0 11.47-14.11 13.48-28.55s.89-6.16.89-6.16Z" style="fill:#004f91" transform="translate(0 -33.12)"/><circle cx="128" cy="94.88" r="76" style="stroke-width:5px;fill:url(#Dégradé_sans_nom_8);stroke:#1d4156;stroke-miterlimit:10"/><circle cx="128" cy="94.88" r="92.5" style="fill:none;stroke-width:4.77px;stroke:#1d4156;stroke-miterlimit:10"/><text style="font-size:32px;fill:#fff;font-family:Arial-BoldMT,Arial;font-weight:700" transform="rotate(-9.55 733.2 -471.458)">EUP<tspan x="65.8" y="0" style="letter-spacing:-.02em">L</tspan></text><path d="m128 27.18 2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM159 36.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM181 59.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM186 84.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM181 112.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM160 134.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM128 143.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM94 134.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM74 112.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71-5.71-4.15-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM69 84.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71L69 97.59l-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM74 57.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71L74 70.59l-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7zM95 35.18l2.18 6.7h7.05l-5.7 4.15 2.18 6.71L95 48.59l-5.71 4.15 2.18-6.71-5.7-4.15h7.05l2.18-6.7z" class="cls-9"/></svg></a>`


Coalispr introduction    
=====================

|prgnam| (*CO*\unt *ALI*\gned *SP*\ecified *R*\eads) is a Python_ tool to clean up (small) RNA sequencing results. It can visualize over 100 :term:`bedgraph`\ s in one panel [#smllgnm]_ and helps to retrieve read counts from associated :term:`bam` files without reliance on reference features (:term:`GTF` annotations).

|

Features
--------

- *Fast and voluminous*
       | Reduced resolution decreases memory-use and speeds up comparison.
       | Handle a large number of samples simultaneously.
       | Count :term:`specified` aligned reads [#allr]_.
       | Count :term:`collapsed` instead of single reads.



.. sidebar:: Bedgraph signals
   
   |
   | :term:`Specific` and :term:`Unspecific` reads. 
 
   .. image:: ../_static/images/specific_light_ed.svg
      :name: logo_smu
      :align: center
      :width: 270 px
      :height: 132 px
      :scale: 80 %
 
   |     
   | :term:`siRNA` and :term:`Reference` RNA.

   .. image:: ../_static/images/reference_light_ed.svg
      :name: logo_ref
      :align: center
      :width: 270 px
      :height: 138 px
      :scale: 80 %




- *Input* files are :term:`bedgraph` files
        | Bedgraph-data are imported for processing with Pandas_.
        | Reads are collected by their mid-point into :term:`bin`\ s [#bin]_.
        | A common index [#idx]_ enables comparison between all samples.
        | Comparisons are done per chromosome [#chr]_ strand.
        | :term:`Specific` reads (:ref:`S, M<logo_smu>`) are separated from :term:`unspecific` reads (:ref:`U<logo_smu>`)
        | (by checking bin-overlap first, then, signal difference).
        | For fast reuse, created datastructures are stored in a binary format.

- *Interactive visualization*
        | All bedgraphs for a chromosome are shown (with Matplotlib_).
        | Toggle data display.
        | Load :term:`GTF` files for overlap with annotated features.
        | Signal scale: normal or log2.
        | Include :term:`reference` RNA sequencing data (:ref:`R<logo_ref>`).
        | Save snapshots as svg, jpg, pdf or png.

- *Counting*
        | Map contiguous regions of :term:`unspecific` or :term:`specific` reads.
        | These segment definitions are stored in :term:`tsv` files to: 
        | Retrieve :term:`specified` reads from :term:`bam` files with Pysam_.
        | Split segments into a number of bins to profile coverage.
        | Collect counts for various read properties and save to tsv files.
        | Obtain counts for particular chromosomal regions.
        | Thus, counting relies on genome coordinates, not GTF references.


.. sidebar:: Counts analysis
   
   |
   | :term:`Specified` reads. 
 
   .. image:: ../_static/images/counts_light_.svg
      :name: cnts_smu
      :align: center
      :width: 270 px
      :height: 150 px
      :scale: 80 %


- *Analysis*
        | Count-outputs can be diagrammed (with Matplotlib_ and Seaborn_).
        | Compare numbers for reads, cDNAs, introns, multimappers.
        | Check length-distributions of reads, also for a particular genomic region.
        | Annotate count files with gene-information from GTF references.
        |

For a rationale and application of |prgnam| see the :doc:`essay <paper>`: '\ |papertitl|'.

|

Requirements
------------
- Preparation (used in the normal work flow and in the :doc:`/tutorials`):
    Bash_, Flexbar_, pyCRAC_, Samtools_, SRA-toolkit_, STAR_ (or another aligner).
- Run-time:
    Python_, Numpy_, Pandas_, Matplotlib_, Pysam_, Seaborn_. [#verss]_
- Enough RAM for loading genome data.

The numerical expression evaluator for NumPy, Numexpr_, can help to get the most of your machine computing capabilities [#pandasdep]_.

|

Installation
------------
|prgnam| is on https://codeberg.org/coalispr/coalispr and pypi.org_ from where it can be downloaded. 

..  pypi.org_ 

Configuration files with properties will have to be edited by the user to analyze their own data (see :doc:`/tutorials`).
Therefore, this package is best installed locally in user space, not system wide. Alternatively, the program can be installed in a virtual environment [#virt]_.

After extraction of the source archive, go to the ``coalispr`` project folder with the ``setup.py`` and ``pyproject.toml`` files and run in a terminal (as user):

.. code-block:: bash

	python3 -m pip install --editable .

This also makes it easy to adapt source code and directly test the changes. 

A script, callable from the command line with ``coalispr``, will be installed locally [#loca]_ (alternatively, you can run ``python3 -m coalispr`` instead of ``coalispr``).

With installs of ``pandas-2.x`` please link ``coalispr/resources/numeric.py`` to ``python3/site_packages/pandas/core/indexes/`` (see :ref:`here <pnds2>`).

Installation can be done in a virtual environment as described in INSTALL.txt




|

Run |prgnam|
------------
In a terminal run the following command-line, which shows the various options for |prgnam|:

.. code-block:: bash

        coalispr -h

See the :doc:`/howtoguides` and the :doc:`/tutorials` for more information.

|

Contribute
----------
All resources for |prgnam| are accessible at\ |nbsp|\ |nbsp|\ |codebergicon|\ |nbsp| Codeberg.org.

- Source Code: https://codeberg.org/coalispr/coalispr
- Issue Tracker: https://codeberg.org/coalispr/coalispr/issues

|

Documentation
-------------
- This documentation is online at https://coalispr.codeberg.page/
- Sources for the documentation can be found at https://codeberg.org/coalispr/coalispr/docs
- Datasets supporting the tutorials are published at `Zenodo.org <https://zenodo.org>`_ under `DOI 10.5281/zenodo.12822543`_

|

Licences
--------
- The program source code is published under the European Union Public Licence (EUPL_)\ |nbsp|\ |nbsp|\  |euplsvg|
- The documentation is under the Creative Commons Attribution License (`CC-BY-4.0 <CCBY4_>`_)\ |nbsp|\ |nbsp|\ |ccby4|

|

Author
------
- The author has been trained as a molecular biologist |orcidicon| and from that angle got involved with high-throughput analysis (see :doc:`background`).

|
    
=====

Notes
'''''

.. [#smllgnm] This has been tested for siRNA sequences mapped to a relatively small fungal genome (~19 Mbp).
.. [#allr] In separate runs, one for :term:`specific` and the other for :term:`unspecific` reads, all count information can be collected.
.. [#idx] The common index is described in the ``genom`` :ref:`module <dointrvl>`.
.. [#bin] Reads are compared by their coverage (i.e. mapping coordinates on a chromosome, with the midpoint halfway the spanned section). The bedgraph values of reads that are placed in a bin are summed. The length of the region spanned by a bin can be set in a configuration file.
.. [#chr] Chromosomes differ in length and cannot be easily compared against each other.
.. [#verss] Versions used for testing and developing |prgnam| are in the ``pyproject.toml``
.. [#pandasdep] See pandas-docs_ for more information and other dependencies.
.. [#virt] See the ``INSTALL.txt`` in the |prgnam| source folder for how to set up a virtual environment. This suggests to extract the |prgnam| sources to the virtual environment folder. An advantage is that the environment will not be affected by other program versions and can be managed by the user (not root). A draw-back is that common python programs get duplicated.

.. rst-class:: asfootnote

.. code-block:: bash

    # cd to folder with virtual environments
    # create environment 1 (env1) with module venv
        bash-5.2$ python3 -m venv env1
    # activate env1
        bash-5.2$ source env1/bin/activate
        (env1) bash-5.2$       
    # extract dist/package
        (env1) bash-5.2$ tar -xvzf /<path_to>/coalispr-$VERSION.tar.gz
        (env1) bash-5.2$ cd coalispr-$VERSION
    # install with:
        (env1) bash-5.2$ python3 -m pip install --editable .
    # add link when using pandas-2.x
        (env1) bash-5.2$ ln -s -r coalispr-$VERSION/coalispr/resources/numeric.py \
                        -t env1/lib/python3.11/site-packages/pandas/core/indexes/
    # run:
        (env1) bash-5.2$ coalispr -h
    # stop the virtual environment:
        (env1) bash-5.2$ deactivate
         bash-5.2$
   
.. [#loca] On a linux system, the script will be placed in ``~/.local/bin/``, which needs to be on the ``$PATH`` to be found when calling it from the command line. A way to accomplish this is by adding the line ``export PATH=$PATH:~/.local/bin`` to the ``~/.bashrc`` or ``~/.bash_aliases``. In a virtual environment [#virt]_, the script ends up in ``/env1/bin/``.


